lib/auth/iam_join.go

```diff
@@ -0,0 +1,301 @@
+// and "?" to match any single character, see https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_resource.html
+func arnMatches(pattern, arn string) (bool, error) {
+	// asterisk should match zero or
+	pattern = strings.ReplaceAll(pattern, "*", ".*")
+	pattern = strings.ReplaceAll(pattern, "?", ".")
+	pattern = "^" + pattern + "$"
+	matched, err := regexp.MatchString(pattern, arn)
+	return matched, trace.Wrap(err)
+}
+
+func checkIAMAllowRules(identity awsIdentity, provisionToken types.ProvisionToken) error {
+	allowRules := provisionToken.GetAllowRules()
+	for _, rule := range allowRules {
+		// If this rule specifies an AWS account, the identity must match
+		if len(rule.AWSAccount) > 0 {
+			if rule.AWSAccount != identity.Account {
+				// account doesn't match, continue to check the next rule
+				continue
+			}
+		}
+		// If this rule specifies an AWS ARN, the identity must match
+		if len(rule.AWSARN) > 0 {
+			matches, err := arnMatches(rule.AWSARN, identity.Arn)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			if !matches {
+				// arn doesn't match, continue to check the next rule
+				continue
+			}
+		}
+		// node identity matches this allow rule
+		return nil
+	}
+	return trace.AccessDenied("instance did not match any allow rules")
+}
+
+func (a *Server) checkIAMRequest(ctx context.Context, client stsClient, challenge string, req *types.RegisterUsingTokenRequest) error {
+	tokenName := req.Token
+	provisionToken, err := a.GetCache().GetToken(ctx, tokenName)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if provisionToken.GetJoinMethod() != types.JoinMethodIAM {
+		return trace.AccessDenied("this token does not support the IAM join method")
+	}
+
+	// parse the incoming http request to the sts:GetCallerIdentity endpoint
+	identityRequest, err := parseSTSRequest(req.STSIdentityRequest)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// validate that the host, method, and headers are correct and the expected
+	// challenge is included in the signed portion of the request
+	if err := validateSTSIdentityRequest(identityRequest, challenge); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// send the signed request to the public AWS API and get the node identity
+	// from the response
+	identity, err := executeSTSIdentityRequest(ctx, client, identityRequest)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// check that the node identity matches an allow rule for this token
+	if err := checkIAMAllowRules(identity, provisionToken); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (a *Server) registerUsingIAMMethod(srv proto.AuthService_RegisterUsingIAMMethodServer) error {
+	ctx := srv.Context()
+
+	p, ok := peer.FromContext(ctx)
+	if !ok {
+		return trace.AccessDenied("failed to read peer information from gRPC context")
+	}
+	remoteAddr := p.Addr.String()
+
+	// read 32 crypto-random bytes to generate the challenge
```

Comment:
Might want to extract the challenge generation code out into its own function - it feels like a digression from the rest of the flow.